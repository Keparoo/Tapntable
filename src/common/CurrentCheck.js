import React, { useState } from 'react';
import moment from 'moment';
import TapntableApi from '../api/api';
import { clearCurrentCheck } from '../actions/currentCheck';
import { v4 as uuid } from 'uuid';
import { useSelector, useDispatch } from 'react-redux';
import { Typography, Container, Button, Stack } from '@mui/material';

const CurrentCheck = ({ orderCatsOn, reload }) => {
  console.debug('CurrentCheck');

  const check = useSelector((st) => st.currentCheck);
  const dispatch = useDispatch();

  const [ showPayScreen, setShowPayScreen ] = useState(false);

  // console.log('Check items', check.items, check.createdAt);

  const sendOrder = async () => {
    console.debug('sendOrder');

    // Create order in db: get order id
    // Hard code userId=1
    const order = await TapntableApi.createOrder(1);
    console.log('order', order, check.newItems);

    if (check.id) {
      // Create ordered_items objects for each item
      // Hard-code seat-num=1
      // Hard-code itemNote="Well Done"
      for (const item of check.newItems) {
        const ordItem = await TapntableApi.createOrdItem(
          item.id,
          order.id,
          check.id,
          1,
          'Well Done'
        );
        console.log('ordItem', ordItem);
      }
    } else {
      // Create Check in db, get Check Id,
      // hard code userId=1, ignore customer field
      const checkRes = await TapntableApi.createCheck(
        1,
        check.tableNum,
        check.numGuests
      );
      // console.log('check', checkRes);

      // Create ordered_items objects for each item
      // Hard-code seat-num=1
      // Hard-code itemNote="Well Done"
      for (const item of check.newItems) {
        const ordItem = await TapntableApi.createOrdItem(
          item.id,
          order.id,
          checkRes.id,
          1,
          'Well Done'
        );
        console.log('ordItem', ordItem);
      }
    }

    // Clear current check
    dispatch(clearCurrentCheck());
    // Return to server page (show open checks)
    reload(true);
    orderCatsOn(false);
    console.log('reload=true');
  };

  // Go back to OpenCheck
  const cancel = () => {
    dispatch(clearCurrentCheck());
    orderCatsOn(false);
  };

  const pay = () => {
    //Show Pay Screen
    orderCatsOn(false);
    // Get type, Amount
    // insert into payments
    // add all payments
    //if totalpayments  >= total: allow close check
  };

  const renderCurrentCheck = () => {
    return (
      <Container sx={{ padding: '8px' }}>
        <div
          style={{ background: 'lightgray', height: '80vh', padding: '8px' }}
        >
          <Typography variant="h5" align="center">
            Current Check
          </Typography>

          <Typography variant="p">
            {check.tableNum && (
              <span>
                Table Num: <strong>{check.tableNum}</strong>
              </span>
            )}
            {check.numGuests && (
              <span style={{ float: 'right' }}>
                Num Guests: <strong>{check.numGuests}</strong>
              </span>
            )}
          </Typography>
          <br />
          <br />

          {check.items.map((i) => (
            <p key={uuid()}>
              <strong>{i.name}</strong>{' '}
              <span style={{ float: 'right' }}>${i.price}</span>
            </p>
          ))}
          {check.newItems.map((i) => (
            <p key={uuid()}>
              <strong>{i.name}</strong>{' '}
              <span style={{ float: 'right' }}>${i.price}</span>
            </p>
          ))}
        </div>

        <div>
          <Typography variant="p" sx={{ padding: '6px' }}>
            {check.createdAt && (
              <span>
                Created At:{' '}
                <strong>{moment(check.createdAt).format('LT')}</strong>
              </span>
            )}
            {(check.subtotal || check.subtotal === 0) && (
              <span style={{ float: 'right', paddingRight: '8px' }}>
                Subtotal: <strong>${check.subtotal.toFixed(2)}</strong>
              </span>
            )}
          </Typography>
          <br />

          <br />
          <Stack direction="row" spacing={2} justifyContent="center">
            <Button onClick={sendOrder} variant="contained">
              Send Order
            </Button>
            <Button onClick={cancel} color="warning" variant="contained">
              Cancel
            </Button>
            <Button onClick={pay} variant="contained">
              Pay
            </Button>
          </Stack>
        </div>
      </Container>
    );
  };

  return renderCurrentCheck();
};

export default CurrentCheck;
