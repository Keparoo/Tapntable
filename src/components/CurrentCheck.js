import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import {
  clearCurrentCheck,
  removeItemFromCheck
} from '../actions/currentCheck';
import sendOrder from '../utils/sendOrder';
import TapntableApi from '../api/api';

import ItemNoteForm from './ItemNoteForm';

import { v4 as uuid } from 'uuid';
import moment from 'moment';

import {
  Typography,
  Container,
  Button,
  Box,
  Stack,
  Divider,
  List,
  ListItem,
  ListItemText,
  IconButton
} from '@mui/material';
import DeleteIcon from '@mui/icons-material/Delete';
import { fetchItemsFromAPI } from '../actions/items';
// import MoreVertIcon from '@mui/icons-material/MoreVert';

const CurrentCheck = ({ showOrderCats, reload, showPayment }) => {
  console.debug('CurrentCheck');

  const dispatch = useDispatch();
  const history = useHistory();
  const [ showItemNoteForm, setShowItemNoteForm ] = useState(false);
  const [ currItem, setCurrItem ] = useState({});

  // Get current user and check
  const user = useSelector((st) => st.user, shallowEqual);
  const check = useSelector((st) => st.currentCheck, shallowEqual);
  console.debug('CurrentCheck', check);

  const sendAndClear = () => {
    sendOrder(check, user, reload, showOrderCats);
    // Clear current check
    dispatch(clearCurrentCheck());
    // Return to server page (show open checks)
    reload(true);
    showOrderCats(false);
  };

  // Go back to OpenCheck
  const cancel = () => {
    dispatch(clearCurrentCheck());
    showOrderCats(false);
    showPayment(false);
    history.push('/');
  };

  // Show Pay Screen
  const pay = () => {
    showOrderCats(false);
    showPayment(true);
  };

  // Calculate and Print Check
  const printCheck = async () => {
    const printCheck = await TapntableApi.printCheck(
      check.id,
      check.subtotal,
      check.localTax,
      check.stateTax,
      check.federalTax
    );
    console.log('printCheck', printCheck);
    dispatch(clearCurrentCheck());
    reload(true);
    showOrderCats(false);
    // Insert logic to print at local printer when available
  };

  // Add an item note: Show form
  const addNote = (arr, idx) => {
    console.debug('addNote', idx, arr);
    setShowItemNoteForm(true);
    setCurrItem(idx);
  };

  // Save item note: Hide form
  const saveNote = (i, note) => {
    console.debug('saveNote', i, note);
    i.itemNote = note.note;
    console.debug('Item Note: ', note.note);
    setShowItemNoteForm(false);
  };

  // Cancel item note: Hide form
  const cancelNote = () => {
    console.debug('cancelNote');
    setShowItemNoteForm(false);
  };

  // Delete item from newItems
  const removeItem = (arr, idx) => {
    console.debug('removeItem', arr, idx);
    dispatch(removeItemFromCheck({ arr, idx }));
    // Refetch items to update count on buttons
    dispatch(fetchItemsFromAPI());
  };

  const renderCurrentCheck = () => {
    return (
      <Container sx={{ padding: '8px', width: 400, marginRight: 0 }}>
        <Box
          px={1}
          py={2}
          sx={{
            height: '80vh',
            width: '100%',
            maxWidth: 350,
            minWidth: 300,
            marginRight: 0,
            bgcolor: 'lightgray'
          }}
          style={{ overflow: 'auto' }}
        >
          <header className="CurrentCheck-Header">
            <Typography variant="h6" align="center">
              {check.tableNum && (
                <span>
                  Table: <strong>{check.tableNum}</strong>
                </span>
              )}
            </Typography>

            <Typography>
              {check.createdAt && (
                <span>
                  Created At:{' '}
                  <strong>{moment(check.createdAt).format('LT')}</strong>
                </span>
              )}
              {check.numGuests && (
                <span style={{ float: 'right' }}>
                  Num Guests: <strong>{check.numGuests}</strong>
                </span>
              )}
            </Typography>
          </header>

          {check.items.length !== 0 && <Divider>Sent Items</Divider>}

          <List className="CurrentCheck-SentItems" dense={true}>
            {check.items.map((i) => (
              <ListItem key={uuid()}>
                <ListItemText
                  primary={
                    <React.Fragment key={uuid()}>
                      <strong>{i.name}</strong>{' '}
                      <span style={{ float: 'right' }}>${i.price}</span>
                    </React.Fragment>
                  }
                  secondary={
                    <React.Fragment key={uuid()}>
                      {i.mods.length !== 0 && (
                        <List component="span">
                          {i.mods.map((m) => (
                            <ListItem
                              sx={{
                                display: 'inline',
                                marginLeft: '1.3em'
                              }}
                              variant="body2"
                              color="text.secondary"
                              key={uuid()}
                              component="span"
                            >
                              {m.modName}
                              {m.modPrice && (
                                <span style={{ float: 'right' }}>
                                  ${m.modPrice}
                                </span>
                              )}
                              <br />
                            </ListItem>
                          ))}
                          {i.itemNote && (
                            <ListItem
                              sx={{
                                display: 'inline',
                                marginLeft: '1.3em'
                              }}
                              variant="body2"
                              color="text.secondary"
                              key={uuid()}
                              component="span"
                            >
                              <strong>****{i.itemNote}</strong>
                            </ListItem>
                          )}
                        </List>
                      )}
                    </React.Fragment>
                  }
                />
              </ListItem>
            ))}
          </List>
          {check.newItems.length !== 0 && <Divider>New Items</Divider>}

          <List className="CurrentCheck-NewItems" dense={true}>
            {check.newItems.map((i, idx, arr) => (
              <React.Fragment key={uuid()}>
                <ListItem key={uuid()} alignItems="flex-start">
                  <IconButton key={uuid()} onClick={() => removeItem(arr, idx)}>
                    <DeleteIcon />
                  </IconButton>
                  <ListItemText
                    onClick={() => addNote(arr, idx)}
                    primary={
                      <React.Fragment>
                        <strong>{i.name}</strong>{' '}
                        <span style={{ float: 'right' }}>${i.price}</span>
                      </React.Fragment>
                    }
                    secondary={
                      <React.Fragment key={uuid()}>
                        {(i.mods.length !== 0 || i.itemNote) && (
                          <List component="span">
                            {i.mods.map((m) => (
                              <ListItem
                                sx={{
                                  display: 'inline',
                                  marginLeft: '1.3em'
                                }}
                                variant="body2"
                                color="text.secondary"
                                key={uuid()}
                                component="span"
                              >
                                {m.modName}
                                {m.modPrice && (
                                  <span style={{ float: 'right' }}>
                                    ${m.modPrice}
                                  </span>
                                )}
                                <br />
                              </ListItem>
                            ))}
                            {i.itemNote && (
                              <ListItem
                                sx={{
                                  display: 'inline',
                                  marginLeft: '1.3em'
                                }}
                                variant="body2"
                                color="text.secondary"
                                key={uuid()}
                              >
                                <strong>****{i.itemNote}</strong>
                              </ListItem>
                            )}
                          </List>
                        )}
                      </React.Fragment>
                    }
                  />
                </ListItem>
                <ListItem>
                  {showItemNoteForm &&
                  currItem === idx && (
                    <ItemNoteForm
                      i={i}
                      idx={idx}
                      save={saveNote}
                      cancel={cancelNote}
                    />
                  )}
                </ListItem>
              </React.Fragment>
            ))}
          </List>
        </Box>

        <footer style={{ marginTop: 'auto' }}>
          <div
            className="CurrentCheck-Payments"
            style={{ position: 'absolute', bottom: '13em' }}
          >
            {check.payments.map((p) => (
              <p key={uuid()} style={{ dispaly: 'inline' }}>
                Payment------{p.type}----{' '}
                <span style={{ float: 'right' }}>${p.subtotal}</span>
              </p>
            ))}
          </div>

          <div className="CurrentCheck-Totals">
            <Typography variant="p" sx={{ padding: '6px' }}>
              {(check.subtotal || check.subtotal === 0) && (
                <span style={{ float: 'right', paddingRight: '16px' }}>
                  Subtotal: <strong>${check.subtotal.toFixed(2)}</strong>
                </span>
              )}
            </Typography>

            <br />
            <Typography variant="p" sx={{ padding: '6px' }}>
              {(check.amountDue || check.amountDue === 0) && (
                <span style={{ float: 'right', paddingRight: '16px' }}>
                  Amount Due: <strong>${check.amountDue.toFixed(2)}</strong>
                </span>
              )}
            </Typography>
            <Typography variant="p" sx={{ padding: '6px' }}>
              {(check.totalTax || check.totalTax === 0) && (
                <span style={{ float: 'left', paddingLeft: '6px' }}>
                  Tax: <strong>${check.totalTax.toFixed(2)}</strong>
                </span>
              )}
            </Typography>
            <br />
          </div>
        </footer>

        <div className="CurrentCheck-Buttons">
          <Stack direction="row" spacing={2} justifyContent="center">
            {check.newItems.length === 0 ? (
              <Button onClick={sendOrder} variant="contained" disabled>
                Send Order
              </Button>
            ) : (
              <Button onClick={sendAndClear} variant="contained">
                Send Order
              </Button>
            )}
            <Button onClick={cancel} color="secondary" variant="contained">
              Cancel
            </Button>
            {check.id ? (
              <Button onClick={printCheck} variant="contained">
                Print Check
              </Button>
            ) : (
              <Button onClick={printCheck} variant="contained" disabled>
                Print Check
              </Button>
            )}
            {check.id ? (
              <Button onClick={pay} variant="contained">
                Pay
              </Button>
            ) : (
              <Button onClick={pay} variant="contained" disabled>
                Pay
              </Button>
            )}
          </Stack>
        </div>
      </Container>
    );
  };

  return renderCurrentCheck();
};

export default CurrentCheck;
