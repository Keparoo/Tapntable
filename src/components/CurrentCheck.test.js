import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import rootReducer from '../reducers/root';
import thunk from 'redux-thunk';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { MemoryRouter } from 'react-router-dom';
import CurrentCheck from './CurrentCheck';

const store = createStore(
  rootReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

it('renders without crashing', function() {
  Element.prototype.scrollIntoView = jest.fn();
  render(
    <Provider store={store}>
      <MemoryRouter>
        <CurrentCheck />
      </MemoryRouter>
    </Provider>
  );
});

it('matches snapshot', function() {
  Element.prototype.scrollIntoView = jest.fn();
  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <CurrentCheck />
      </MemoryRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
