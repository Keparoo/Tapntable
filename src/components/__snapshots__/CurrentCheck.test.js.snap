// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-fgc9rm-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-6zxm88"
      style="overflow: auto;"
    >
      <header
        class="CurrentCheck-Header"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-19z22rr-MuiTypography-root"
        />
      </header>
      <ul
        class="MuiList-root MuiList-dense SentItems css-1mk9mw3-MuiList-root"
      />
      <ul
        class="MuiList-root MuiList-dense SentItems css-1mk9mw3-MuiList-root"
      />
      <ul
        class="MuiList-root MuiList-dense SentItems css-1mk9mw3-MuiList-root"
      />
      <ul
        class="MuiList-root MuiList-dense CurrentCheck-NewItems css-1mk9mw3-MuiList-root"
      />
      <div />
    </div>
    <footer>
      <div
        class="CurrentCheck-Totals"
      >
        <div
          class="CurrentCheck-Payments"
        />
        <span
          class="MuiTypography-root MuiTypography-p css-409jjv-MuiTypography-root"
        >
          <span
            style="float: right; padding-right: 16px;"
          >
            Subtotal: 
            <strong>
              $0.00
            </strong>
          </span>
        </span>
        <br />
        <span
          class="MuiTypography-root MuiTypography-p css-409jjv-MuiTypography-root"
        />
        <span
          class="MuiTypography-root MuiTypography-p css-409jjv-MuiTypography-root"
        />
        <br />
      </div>
    </footer>
    <div
      class="CurrentCheck-Buttons"
    >
      <div
        class="css-1clyo4s-MuiStack-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Send Order
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Print Check
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Pay
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
